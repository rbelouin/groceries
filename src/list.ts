import type { getAllRecipesByName as GetAllRecipesByName, resizeRecipe as ResizeRecipe, Recipe } from "./recipe";
import type { reduceQuantities as ReduceQuantities } from "./quantity";
import type { getStoreArticles as GetStoreArticles, StoreArticle } from "./stores";

export type List = ListItem[];
export type ListItem = {
  name: string;
  quantity: Quantity;
}
export type Quantity = number | string;

export type GeneratedList = Record<ListItem["name"], GeneratedListItem>;
export type GeneratedListItem = {
  quantity: string;
  checked: boolean;
};

export type SortedGeneratedList = { name: string; quantity: string; checked: boolean }[];

declare const getAllRecipesByName: typeof GetAllRecipesByName;
declare const resizeRecipe: typeof ResizeRecipe;
declare const reduceQuantities: typeof ReduceQuantities;
declare const getStoreArticles: typeof GetStoreArticles;
declare const LIST_SHEET_NAME: string;
declare const LIST_SHEET_ARTICLE_RANGE: string;
declare const GENERATED_LIST_SHEET_NAME: string;
declare const GENERATED_LIST_SHEET_ARTICLE_RANGE: string;

export function calculateAndUpdateGeneratedList(spreadsheet: GoogleAppsScript.Spreadsheet.Spreadsheet) {
  const recipesByName = getAllRecipesByName(spreadsheet);
  const articlesByName = getStoreArticles(spreadsheet);

  return updateGeneratedList(spreadsheet, articlesByName, calculateGeneratedList(recipesByName, readList(spreadsheet)));
}

export function readList(spreadsheet: GoogleAppsScript.Spreadsheet.Spreadsheet): List {
  const range = spreadsheet.getRange(`${LIST_SHEET_NAME}!${LIST_SHEET_ARTICLE_RANGE}`);
  return range.getValues().flatMap(([name, quantity]) => name ? [{ name, quantity }] : []);
}

export function calculateGeneratedList(recipesByName: Record<string, Recipe>, list: List): GeneratedList {
  return list
    .flatMap(item => recipesByName[item.name]
      ? resizeRecipe(recipesByName[item.name], item.quantity as `${number}p`).ingredients
      : [item])
    .reduce((acc, item) => ({
      ...acc,
      [item.name]: {
        quantity: reduceQuantities(acc[item.name]?.quantity || "", item.quantity),
        checked: false,
      },
    }), {});
}

export function updateGeneratedList(spreadsheet: GoogleAppsScript.Spreadsheet.Spreadsheet, articlesByName: Record<string, StoreArticle>, list: GeneratedList) {
  const range = spreadsheet.getRange(`${GENERATED_LIST_SHEET_NAME}!${GENERATED_LIST_SHEET_ARTICLE_RANGE}`);

  const oldList = readGeneratedList(range);
  const newList = Object.fromEntries(Object.entries(list).map(([name, item]) => ([name, {
    ...item,
    checked: oldList[name] && oldList[name].quantity === item.quantity ? oldList[name].checked : false,
  }])));

  writeGeneratedList(range, sortGeneratedList(articlesByName, newList));
}

function sortGeneratedList(articlesByName: Record<string, StoreArticle>, list: GeneratedList): SortedGeneratedList {
  const items = Object.entries(list).map(([name, { quantity, checked }]) => ({ checked, name, quantity }));
  items.sort((itemA, itemB) => {
    if (articlesByName[itemA.name] && articlesByName[itemB.name]) {
      if (articlesByName[itemA.name].department > articlesByName[itemB.name].department) {
        return 1;
      }

      if (articlesByName[itemA.name].department < articlesByName[itemB.name].department) {
        return -1;
      }

      return itemA.name > itemB.name ? 1 : -1;
    }

    if (articlesByName[itemA.name] && !articlesByName[itemB.name]) {
      return 1;
    }

    if (!articlesByName[itemA.name] && articlesByName[itemB.name]) {
      return -1;
    }

    return itemA.name > itemB.name ? 1 : -1;
  });

  return items;
}

function readGeneratedList(range: GoogleAppsScript.Spreadsheet.Range): GeneratedList {
  return range.getValues().reduce((acc, [checked, name, quantity]) => !name ? acc : ({
    ...acc,
    [name]: { quantity, checked },
  }), {});
}

function writeGeneratedList(range: GoogleAppsScript.Spreadsheet.Range, list: SortedGeneratedList) {
  const numRows = range.getNumRows();
  const items = list.map(({ name, quantity, checked }) => ([checked, name, quantity]));
  const blankRows = new Array(numRows - items.length).fill(["", "", ""]);

  range.setValues(items.concat(blankRows));
}
