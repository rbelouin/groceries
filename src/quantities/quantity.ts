import { Mass } from "./mass";
import { Volume } from "./volume";
import { Length } from "./length";
import { Area } from "./xarea";
import { PhysicalQuantity } from "./types";

export type QuantityType = (typeof Quantity.prototype.q)["type"];
export type QuantityConversions = Map<String, Map<String, [Quantity, Quantity]>>;

export class Quantity implements PhysicalQuantity {
  q:
    | {
        type: "volume";
        value: Volume;
      }
    | {
        type: "mass";
        value: Mass;
      }
    | {
        type: "length";
        value: Length;
      }
    | {
        type: "area";
        value: Area;
      }
    | {
        type: "unknown";
        unit: string;
        count: number;
      };

  conversions: QuantityConversions;

  constructor(
    q: typeof Quantity.prototype.q,
    conversions?: QuantityConversions,
  ) {
    this.q = q;
    this.conversions = conversions || new Map();
  }

  static parse(
    value: number | string = "",
    conversions: string = "",
  ): Quantity | undefined {
    const parsedConversions = Quantity.parseConversions(conversions);

    if (typeof value === "number") {
      return Quantity.from(value, "", parsedConversions);
    }

    const result = value.match(/^([0-9.]+)\s*(\S*)$/);
    if (!result) return undefined;

    return Quantity.from(parseFloat(result[1]), result[2], parsedConversions);
  }

  static parseConversions(conversions: string = ""): QuantityConversions {
    const acc = new Map();

    conversions.split("\n").forEach((conversion) => {
      if (!conversion) return;

      const [quantity1, quantity2] = Quantity.parseConversion(conversion);
      const innerMap = acc.get(quantity1.getUnitKey()) ?? new Map();
      innerMap.set(quantity2.getUnitKey(), [quantity1, quantity2]);
      acc.set(quantity1.getUnitKey(), innerMap);
    });

    return acc;
  }

  static parseConversion(conversion: string = ""): [Quantity, Quantity] {
    const result = conversion.split("/");
    if (result.length !== 2) {
      throw new Error(`Invalid conversion rule: ${conversion}`);
    }

    const [quantity1, quantity2] = result.map((quantity) => {
      const parsed = Quantity.parse(quantity);
      if (!parsed) {
        throw new Error(`Invalid quantity: ${quantity}`);
      }
      return parsed;
    });

    return [quantity1, quantity2];
  }

  static from(
    count: number,
    unit: string = "",
    conversions?: QuantityConversions,
  ): Quantity {
    if (Volume.supportsUnit(unit)) {
      return new Quantity(
        {
          type: "volume",
          value: Volume.from(count, unit),
        },
        conversions,
      );
    }

    if (Mass.supportsUnit(unit)) {
      return new Quantity(
        {
          type: "mass",
          value: Mass.from(count, unit),
        },
        conversions,
      );
    }

    if (Length.supportsUnit(unit)) {
      return new Quantity(
        {
          type: "length",
          value: Length.from(count, unit),
        },
        conversions,
      );
    }

    if (Area.supportsUnit(unit)) {
      return new Quantity(
        {
          type: "area",
          value: Area.from(count, unit),
        },
        conversions,
      );
    }

    const trimmedUnit = unit.trim();
    if (["", "p"].indexOf(trimmedUnit) < 0) {
      console.warn(`Unrecognized unit: ${trimmedUnit}`);
    }

    return new Quantity(
      {
        type: "unknown",
        unit: trimmedUnit,
        count,
      },
      conversions,
    );
  }

  tryConvertTo(quantity: Quantity): Quantity {
    const rule1 = this.conversions.get(this.getUnitKey())?.get(quantity.getUnitKey());
    if (rule1) {
      const [quantity1, quantity2] = rule1;
      return new Quantity(quantity2.q, this.conversions).multiply(this.divide(quantity1));
    }

    const rule2 = this.conversions.get(quantity.getUnitKey())?.get(this.getUnitKey());
    if (rule2) {
      const [quantity2, quantity1] = rule2;
      return new Quantity(quantity2.q, this.conversions).multiply(this.divide(quantity1));
    }

    return this;
  }

  private getUnitKey(): string {
    return this.q.type === "unknown" ? this.q.unit : this.q.type;
  }

  add(quantity?: Quantity): Quantity {
    if (!quantity) return this;

    const convertedQuantity = quantity.tryConvertTo(this);

    switch (this.q.type) {
      case "volume":
        if (convertedQuantity.q.type === "volume") {
          return new Quantity({
            type: "volume",
            value: this.q.value.add(convertedQuantity.q.value),
          }, this.conversions);
        }
        break;
      case "mass":
        if (convertedQuantity.q.type === "mass") {
          return new Quantity({
            type: "mass",
            value: this.q.value.add(convertedQuantity.q.value),
          }, this.conversions);
        }
        break;
      case "length":
        if (convertedQuantity.q.type === "length") {
          return new Quantity({
            type: "length",
            value: this.q.value.add(convertedQuantity.q.value),
          }, this.conversions);
        }
        break;
      case "area":
        if (convertedQuantity.q.type === "area") {
          return new Quantity({
            type: "area",
            value: this.q.value.add(convertedQuantity.q.value),
          }, this.conversions);
        }
        break;
      case "unknown":
        if (convertedQuantity.q.type === "unknown") {
          if (this.q.unit !== convertedQuantity.q.unit) {
            throw new Error(
              `Incompatible units: ${this.q.unit} vs. ${convertedQuantity.q.unit}`,
            );
          }

          return new Quantity({
            type: "unknown",
            unit: this.q.unit,
            count: this.q.count + convertedQuantity.q.count,
          }, this.conversions);
        }
    }

    throw new Error(
      `Incompatible types: ${this.q.type} vs. ${convertedQuantity.q.type}`,
    );
  }

  multiply(factor: number): Quantity {
    switch (this.q.type) {
      case "volume":
        return new Quantity({
          type: this.q.type,
          value: this.q.value.multiply(factor),
        }, this.conversions);
      case "mass":
        return new Quantity({
          type: this.q.type,
          value: this.q.value.multiply(factor),
        }, this.conversions);
      case "length":
        return new Quantity({
          type: this.q.type,
          value: this.q.value.multiply(factor),
        }, this.conversions);
      case "area":
        return new Quantity({
          type: this.q.type,
          value: this.q.value.multiply(factor),
        }, this.conversions);
      case "unknown":
        return new Quantity({
          type: this.q.type,
          unit: this.q.unit,
          count: this.q.count * factor,
        }, this.conversions);
    }
  }

  divide(quantity: Quantity): number {
    const convertedQuantity = quantity.tryConvertTo(this);

    switch (this.q.type) {
      case "volume":
        if (convertedQuantity.q.type === "volume") {
          return this.q.value.divide(convertedQuantity.q.value);
        }
        break;
      case "mass":
        if (convertedQuantity.q.type === "mass") {
          return this.q.value.divide(convertedQuantity.q.value);
        }
        break;
      case "length":
        if (convertedQuantity.q.type === "length") {
          return this.q.value.divide(convertedQuantity.q.value);
        }
        break;
      case "area":
        if (convertedQuantity.q.type === "area") {
          return this.q.value.divide(convertedQuantity.q.value);
        }
        break;
      case "unknown":
        if (convertedQuantity.q.type === "unknown") {
          if (this.q.unit !== convertedQuantity.q.unit) {
            throw new Error(
              `Incompatible units: ${this.q.unit} vs. ${convertedQuantity.q.unit}`,
            );
          }

          return this.q.count / convertedQuantity.q.count;
        }
    }

    throw new Error(
      `Incompatible types: ${this.q.type} vs. ${convertedQuantity.q.type}`,
    );
  }

  toString(): string {
    switch (this.q.type) {
      case "volume":
      case "mass":
      case "length":
      case "area":
        return this.q.value.toString();
      case "unknown":
        return this.q.unit === "" ? this.q.count.toString() : `${this.q.count} ${this.q.unit}`;
    }
  }
}
