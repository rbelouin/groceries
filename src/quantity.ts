export type Quantity = number | string;

type Countable = {
  type: "countable";
  count: number;
  unit: undefined;
};

const volumeUnits = ["ml", "c-à-c", "cl", "c-à-s", "dl", "l"] as const;

type VolumeUnit = (typeof volumeUnits)[number];
type Volume = {
  type: "volume";
  count: number;
  unit: VolumeUnit;
};

const weightUnits = ["mg", "g", "kg"] as const;

type WeightUnit = (typeof weightUnits)[number];
type Weight = {
  type: "weight";
  count: number;
  unit: WeightUnit;
};

type ParsedQuantity = Countable | Volume | Weight;
type ParsedQuantities = Record<ParsedQuantity["type"], number> & {
  unknown?: string[];
};

export function reduceQuantities(acc: string, quantity: Quantity): string {
  const quantities = addQuantity(parseQuantities(acc), parseQuantity(quantity));
  return Object.entries(quantities).flatMap(([key, value]) => {
    switch (key) {
      case "countable":
        return [(value as ParsedQuantities["countable"]).toString()];
      case "volume":
        return [serializeMilliliters(value as ParsedQuantities["volume"])];
      case "weight":
        return [serializeMilligrams(value as ParsedQuantities["weight"])];
      case "unknown":
        return (value as ParsedQuantities["unknown"]);
    }
  })
    .filter(item => item)
    .join("|");
}

function parseQuantities(value: string): ParsedQuantities {
  return value.split("|").reduce((acc, item) => addQuantity(acc, parseQuantity(item)), {} as ParsedQuantities);
}

function addQuantity(quantities: ParsedQuantities, quantity: ParsedQuantity | string): ParsedQuantities {
  if (typeof quantity === "string") {
    return {
      ...quantities,
      unknown: (quantities.unknown || []).concat(quantity),
    };
  }

  switch (quantity.type) {
    case "countable":
      return {
        ...quantities,
        countable: (quantities.countable || 0) + quantity.count,
      };
    case "volume":
      return {
        ...quantities,
        volume: (quantities.volume || 0) + quantity.count,
      };
    case "weight":
      return {
        ...quantities,
        weight: (quantities.weight || 0) + quantity.count,
      };
  }
}

function toMilliliters(count: number, unit: VolumeUnit): number {
  switch (unit) {
    case "ml":
      return count;
    case "c-à-c":
      return count * 5;
    case "cl":
      return count * 10;
    case "c-à-s":
      return count * 15;
    case "dl":
      return count * 100;
    case "l":
      return count * 1000;
  }
}

export function serializeMilliliters(count: number): string {
  if (count >= 1000) {
    return `${Math.ceil(count / 10) / 100}l`;
  }

  if (count >= 100) {
    return `${Math.ceil(count) / 100}dl`;
  }

  if (count >= 10) {
    return `${Math.ceil(count * 10) / 100}cl`;
  }

  return `${Math.ceil(count * 100) / 100}ml`;
}

function toMilligrams(count: number, unit: WeightUnit): number {
  switch (unit) {
    case "mg":
      return count;
    case "g":
      return 1000 * count;
    case "kg":
      return 1000000 * count;
  }
}

export function serializeMilligrams(count: number): string {
  if (count >= 1000000) {
    return `${Math.ceil(100 * count / 1000 / 1000) / 100}kg`;
  }

  if (count >= 1000) {
    return `${Math.ceil(100 * count / 1000) / 100}g`;
  }

  return `${Math.ceil(100 * count) / 100}mg`;
}

function parseQuantity(value: Quantity): string | ParsedQuantity {
  if (typeof value === "number") {
    return { type: "countable", count: value, unit: undefined };
  }

  if (value === "") {
    return "";
  }

  const result = value.match(/^([0-9.]+)\s*(\S*)$/);

  if (result === null) {
    console.warn(`Unrecognized quantity: ${value}.`);
    return value;
  }

  const [_, countString, unit] = result;
  const count = parseFloat(countString);

  if (unit === "") {
    return { type: "countable", count, unit: undefined };
  }

  if (isVolumeUnit(unit)) {
    return { type: "volume", count: toMilliliters(count, unit), unit: "ml" };
  }

  if (isWeightUnit(unit)) {
    return { type: "weight", count: toMilligrams(count, unit), unit: "mg" };
  }

  console.error(`Unrecognized unit: ${unit}.`);
  return value;
}

function isVolumeUnit(unit: string): unit is VolumeUnit {
  return (volumeUnits as readonly string[]).indexOf(unit) >= 0;
}

function isWeightUnit(unit: string): unit is WeightUnit {
  return (weightUnits as readonly string[]).indexOf(unit) >= 0;
}
