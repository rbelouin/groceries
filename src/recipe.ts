export type Recipe = {
  name: string;
  people: `${number}p`;
  ingredients: Ingredient[];
};

export type Ingredient = {
  name: string;
  quantity?: Quantity;
};

export type Quantity = number | `${number} ${string}` | `${number}${string}` | "";

export function sortAndFormatRecipes() {
  const range = getRange();
  sortRecipes(range);
  formatRecipes(range);
}

export function sortRecipes(range: GoogleAppsScript.Spreadsheet.Range) {
  const recipes = readRecipes(range);

  recipes.sort((a, b) => a.name > b.name ? 1 : -1);

  writeRecipes(range, recipes);
}

function readRecipes(range: GoogleAppsScript.Spreadsheet.Range): Recipe[] {
  let recipes: Recipe[] = [];

  for (let row of range.getValues()) {
    if (row.every(cell => cell === '')) {
      break;
    }

    if (row[0] !== '') {
      recipes.push({
        name: row[0],
        people: row[2],
        ingredients: [],
      });
    } else {
      recipes[recipes.length - 1].ingredients.push({
        name: row[1],
        quantity: row[2],
      });
    }
  }

  return recipes;
}

function writeRecipes(range: GoogleAppsScript.Spreadsheet.Range, recipes: Recipe[]) {
  const values = recipes.flatMap(recipe => ([
    [recipe.name, '', recipe.people],
    ...recipe.ingredients.map(ingredient => (['', ingredient.name, ingredient.quantity]))
  ]));

  const blank = new Array(range.getNumRows() - values.length).fill(['', '', '']);

  range.setValues(values.concat(blank));
}

function getRange(): GoogleAppsScript.Spreadsheet.Range {
  const spreadsheet = SpreadsheetApp.getActiveSpreadsheet();
  return spreadsheet.getRange(`${RECIPE_SHEET_NAME}!${RECIPE_SHEET_RANGE}`);
}

export function resizeRecipe(recipe: Recipe, people: `${number}p`): Recipe {
  const ratio = parseInt(people, 10) / parseInt(recipe.people, 10);

  return {
    ...recipe,
    people,
    ingredients: recipe.ingredients.map(ingredient => resizeIngredient(ingredient, ratio)),
  };
}

function resizeIngredient(ingredient: Ingredient, ratio: number): Ingredient {
  if (typeof ingredient.quantity === "undefined") {
    return ingredient;
  }

  if (typeof ingredient.quantity === "number") {
    return {
      ...ingredient,
      quantity: ingredient.quantity * ratio,
    };
  }

  if (typeof ingredient.quantity === "string") {
    return {
      ...ingredient,
      quantity: ingredient.quantity.replace(/^([0-9,.]+)/, (_, num) => {
        return (Math.round(parseInt(num, 10) * ratio * 100) / 100).toString();
      }) as Quantity,
    };
  }

  throw new Error("Unsupported quantity");
}

export function formatRecipes(range: GoogleAppsScript.Spreadsheet.Range) {
  const values = range.getValues();
  const fontSizes = values.map(row => row.map(_ => row[0] ? 16 : 10));
  range.setFontSizes(fontSizes);
}
