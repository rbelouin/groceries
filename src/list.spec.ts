import { readList, calculateGeneratedList, updateGeneratedList, GeneratedList } from "./list";
import { resizeRecipe, type Recipe } from "./recipe";
import { reduceQuantities } from "./quantity";

describe("list", () => {
  beforeEach(() => {
    (global as any).LIST_SHEET_NAME = "LIST_SHEET_NAME";
    (global as any).LIST_SHEET_ARTICLE_RANGE = "LIST_SHEET_ARTICLE_RANGE";
    (global as any).GENERATED_LIST_SHEET_NAME = "GENERATED_LIST_SHEET_NAME";
    (global as any).GENERATED_LIST_SHEET_ARTICLE_RANGE = "GENERATED_LIST_SHEET_ARTICLE_RANGE";
  });

  afterEach(() => {
    delete (global as any).LIST_SHEET_NAME;
    delete (global as any).LIST_SHEET_ARTICLE_RANGE;
    delete (global as any).GENERATED_LIST_SHEET_NAME;
    delete (global as any).GENERATED_LIST_SHEET_ARTICLE_RANGE;
  });

  describe("readList", () => {
    it("should only read rows with non-empty names", () => {
      const values = [
        ["Pain", "300g"],
        ["Eau", ""],
        ["", ""],
        ["Brioche", "1kg"],
      ];

      const spreadsheet = {
        getRange: (notation: string) => {
          expect(notation).toBe("LIST_SHEET_NAME!LIST_SHEET_ARTICLE_RANGE");
          return createRangeFromValues(values);
        },
      } as Partial<GoogleAppsScript.Spreadsheet.Spreadsheet>;

      expect(readList(spreadsheet as any)).toStrictEqual([
        { name: "Pain", quantity: "300g" },
        { name: "Eau", quantity: "" },
        { name: "Brioche", quantity: "1kg" },
      ]);
    });
  });

  describe("calculateGeneratedList", () => {
    const recipesByName = {
      "Tartines": {
        name: "Tartines",
        people: "4p",
        ingredients: [{
          name: "Pain",
          quantity: "100g",
        }, {
          name: "Confiture",
          quantity: "60ml",
        }],
      },
    } as Record<string, Recipe>;

    const list = [
      { name: "Pain", quantity: "300g" },
      { name: "Tartines", quantity: "2p" },
      { name: "Eau", quantity: "" },
      { name: "Brioche", quantity: "1kg" },
      { name: "Pain", quantity: "600g" },
    ];

    beforeEach(() => {
      (global as any).resizeRecipe = resizeRecipe;
      (global as any).reduceQuantities = reduceQuantities;
    });

    afterEach(() => {
      delete (global as any).resizeRecipe;
      delete (global as any).reduceQuantities;
    });

    it("should expand recipes and deduplicate articles", () => {
      const generatedList = calculateGeneratedList(recipesByName, list);
      expect(Object.keys(generatedList)).toStrictEqual(["Pain", "Confiture", "Eau", "Brioche"]);
    });

    it("should scale and combine quantity fields of redundant items", () => {
      const generatedList = calculateGeneratedList(recipesByName, list);
      expect(generatedList["Pain"].quantity).toBe("950g");
    });

    it("should leave the other quantity fields uncombined", () => {
      const generatedList = calculateGeneratedList(recipesByName, list);
      expect(generatedList["Eau"].quantity).toBe("");
      expect(generatedList["Brioche"].quantity).toBe("1kg");
      expect(generatedList["Confiture"].quantity).toBe("3cl");
    });

    it("should mark articles as unchecked", () => {
      const generatedList = calculateGeneratedList(recipesByName, list);
      expect(Object.values(generatedList).map(item => item.checked)).not.toEqual(expect.arrayContaining([true]));
    });
  });

  describe("updateGeneratedList", () => {
    let range: GoogleAppsScript.Spreadsheet.Range;
    let list: GeneratedList;
    let spreadsheet: Partial<GoogleAppsScript.Spreadsheet.Spreadsheet>;

    beforeEach(() => {
      range = createRangeFromValues([
        [false, "Pain", "500g"],
        [true, "Baguette", "250g"],
        [false, "Beurre", "500g"],
        [true, "Crème Fraîche", "5cl"],
        [false, "Yaourt", "6 pots"],
        [true, "Œufs", "6"],
      ]);

      list = {
        "Pain": {
          quantity: "500g",
          checked: false,
        },
        "Baguette": {
          quantity: "250g",
          checked: false,
        },
        "Saucisson": {
          quantity: "100g",
          checked: false,
        },
        "Yaourt": {
          quantity: "1l",
          checked: false,
        },
        "Œufs": {
          quantity: "12",
          checked: false,
        },
      };

      spreadsheet = {
        getRange: (notation: string) => {
          expect(notation).toBe("GENERATED_LIST_SHEET_NAME!GENERATED_LIST_SHEET_ARTICLE_RANGE");
          return range;
        },
      } as Partial<GoogleAppsScript.Spreadsheet.Spreadsheet>;
    });

    it("should keep articles that haven’t changed with the same `checked` status", () => {
      updateGeneratedList(spreadsheet as any, list);

      expect(spreadsheet.getRange!("GENERATED_LIST_SHEET_NAME!GENERATED_LIST_SHEET_ARTICLE_RANGE").getValues()).toStrictEqual(expect.arrayContaining([
        [false, "Pain", "500g"],
        [true, "Baguette", "250g"],
      ]));
    });

    it("should remove articles that are no longer listed", () => {
      updateGeneratedList(spreadsheet as any, list);

      expect(spreadsheet.getRange!("GENERATED_LIST_SHEET_NAME!GENERATED_LIST_SHEET_ARTICLE_RANGE").getValues()).toStrictEqual(expect.not.arrayContaining([[false, "Beurre", "500g"]]));
      expect(spreadsheet.getRange!("GENERATED_LIST_SHEET_NAME!GENERATED_LIST_SHEET_ARTICLE_RANGE").getValues()).toStrictEqual(expect.not.arrayContaining([[true, "Crème Fraîche", "5cl"]]));
    });

    it("should add articles that were not listed before", () => {
      updateGeneratedList(spreadsheet as any, list);

      expect(spreadsheet.getRange!("GENERATED_LIST_SHEET_NAME!GENERATED_LIST_SHEET_ARTICLE_RANGE").getValues()).toStrictEqual(expect.arrayContaining([[false, "Saucisson", "100g"]]));
    });

    it("should 'uncheck' articles that have changed quantities", () => {
      updateGeneratedList(spreadsheet as any, list);

      expect(spreadsheet.getRange!("GENERATED_LIST_SHEET_NAME!GENERATED_LIST_SHEET_ARTICLE_RANGE").getValues()).toStrictEqual(expect.arrayContaining([
        [false, "Yaourt", "1l"],
        [false, "Œufs", "12"],
      ]));
    });

    it("should match snapshot", () => {
      updateGeneratedList(spreadsheet as any, list);

      expect(spreadsheet.getRange!("GENERATED_LIST_SHEET_NAME!GENERATED_LIST_SHEET_ARTICLE_RANGE").getValues()).toMatchSnapshot();
    });
  });
});

function createRangeFromValues(initialValues: any[][]): GoogleAppsScript.Spreadsheet.Range {
  let values = initialValues;

  const range = {
    getNumRows: () => values.length,
    getValues: () => values,
    setValues: (newValues: any[][]) => { values = newValues },
  } as Partial<GoogleAppsScript.Spreadsheet.Range>;

  return range as any;
}
