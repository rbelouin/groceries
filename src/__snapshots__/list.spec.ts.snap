// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list updateGeneratedList should match snapshot 1`] = `
[
  [
    true,
    "Baguette",
    "250g",
  ],
  [
    false,
    "Pain",
    "500g",
  ],
  [
    false,
    "Saucisson",
    "100g",
  ],
  [
    false,
    "Yaourt",
    "1l",
  ],
  [
    false,
    "Å’ufs",
    "12",
  ],
  [
    "",
    "",
    "",
  ],
]
`;
